### ISSUE TITLE
Xamarin.iOS - exception caused by Linker?

### ISSUE THREAD

**Issue by jamesatgithub** [AutoMapper/AutoMapper#1846]
I got below error on real device with "Link SDK Assemblies Only", but no problem with "Don't Link", is there any way to avoid the exception with "Link SDK Assemblies Only"?

Value cannot be null.
Parameter name: method (System.ArgumentNullException)
  at System.Linq.Expressions.Expression.Call (System.Reflection.MethodInfo method, System.Linq.Expressions.Expression arg0) [0x00034] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/dlr/Runtime/Microsoft.Scripting.Core/Ast/MethodCallExpression.cs:526 
  at AutoMapper.Mappers.ConvertMapper.ConvertExpression (System.Type sourceType, System.Type destinationType) [0x0003d] in <628f28fd2f2f4658a1addc54561a6c25>:0 
  at AutoMapper.Mappers.ConvertMapper+<>c__DisplayClass1_1.<GetConverters>b__4 () [0x00000] in <628f28fd2f2f4658a1addc54561a6c25>:0 
  at System.Lazy`1[T].CreateValue () [0x00075] in <15e850188d9f425bbeae90f0bbc51e17>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.T
hrow () [0x0000c] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:143 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00047] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System
.Threading.Tasks.Task task) [0x0000b] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:113

**Comment by lbargaoanu** [comment]
I don't think this approach is useful with AM. Because AM generates code at run time so it's not possible to know at compile time what it needs. Unless, as with .NET Native, you can tell it what you need. But this seems complicated.

**Comment by jamesatgithub** [comment]
@lbargaoanu I'm new to AM, do you mean AM not ready for Xamarin.iOS yet (I used the latest 5.2)? or there is some rules to follow when using with Xamarin.iOS? I didn't see any info in Wiki for iOS specially.

**Comment by lbargaoanu** [comment]
No, there is nothing wrong in AM. Research a little those configuration options and read my answer again. It might make more sense.

**Comment by jamesatgithub** [comment]
I tried below simple mapping configuration to create a mapper, sid is type of string and Id is type of int:
` cfg.CreateMap<ApprovalHQResult, ApprovalSubmitHQInputDto>()
                .ForMember(d => d.sid, opt => opt.MapFrom(s => s.Id))`

it cause below exception with "Link SDK Assemblies Only" but works with "Don't Link" on iOS

System.ArgumentNullException: Value cannot be null.
Parameter name: method
  at System.Dynamic.Utils.ContractUtils.RequiresNotNull (System.Object value, System.String paramName) [0x00006] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/dlr/Runtime/Microsoft.Scripting.Core/Utils/ContractUtils.cs:58 
  at System.Linq.Expressions.Expression.Call (System.Reflection.MethodInfo method, System.Linq.Expressions.Expression arg0) [0x00000] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/dlr/Runtime/Microsoft.Scripting.Core/Ast/MethodCallExpression.cs:517 
  at AutoMapper.Mappers.ConvertMapper.ConvertExpression (System.Type sourceType, System.Type destinationType) [0x0003d] in <628f28fd2f2f4658a1addc54561a6c25>:0 
  at AutoMapper.Mappers.ConvertMapper+<>
c__DisplayClass1_1.<GetConverters>b__4 () [0x00000] in <628f28fd2f2f4658a1addc54561a6c25>:0 
  at System.Lazy`1[T].CreateValue () [0x00075] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/referencesource/mscorlib/system/Lazy.cs:437 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/referencesource/mscorlib/system/runtime/exceptionservices/exceptionservicescommon.cs:143 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x00047] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/referencesource/mscorlib/system/r
untime/compilerservices/TaskAwaiter.cs:187 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:156 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaiter.cs:128 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in /Users/builder/data/lanes/3969/44931ae8/source/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/src/mono/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/TaskAwaite
r.cs:113

**Comment by jamesatgithub** [comment]
@lbargaoanu so anything I can do to use AM with xamarin.ios without above error?

**Comment by lbargaoanu** [comment]
I think you need to ask this in an iOS forum.

**Comment by johndbell** [comment]
@jamesatgithub did you get any resolution to the ArgumentNullException at System.Dynamic.Utils.ContractUtils.RequiresNotNull? I'm seeing the same thing and can't find a solution.

**Comment by jamesatgithub** [comment]
@johndbell not yet. I give up using AutoMapper in my last iOS app. Do you still see this error with latest Xamarin?

**Comment by hvaughan3** [comment]
@jamesatgithub  @johndbell I ran across the same issue just now. I noticed that the issue did not happen when I used my Debug or Release configuration but it did happen when I used 2 different custom configurations I had created long before.

The weird thing was that I thought that this had worked before. Anyway, I ended up copying over my iOS project's .csproj settings from the `Debug|iPhoneSimulator`'s `PropertyGroup` section into my custom configuration's `PropertyGroup` section and everything started working again. The same thing happened for both configurations. I also deleted all **obj** and **bin** folders and deleted the app from the simulator after doing this.

I tried going back and seeing what properties made the difference, moved properties around but started running into another reflection related exception with System.Net.HttpClient... very strange stuff. At that point I had wasted half my day and said to hell with it.

**Comment by jamesatgithub** [comment]
@hvaughan3 I still think this is due to the Linker in Xamarin.iOS, there should be a document for AM about how to deal with linker for Release mode ("Link SDK Assemblies Only" or “Link All"). I didn't use AM for Xamarin.iOS any more since my last try, but still be interested. I will try when get time. Keep me updated if you find any solution.

some references about linker:
https://lostechies.com/jimmybogard/2014/11/11/dealing-with-the-linker-in-xamarin-apps/
https://forums.xamarin.com/discussion/35911/automapper-linker-issue

**Comment by lbargaoanu** [comment]
Feel free to add smth in the wiki.

**Comment by hvaughan3** [comment]
@jamesatgithub While trying to push out a new update to the stores, I was not even able to build the app, talking about not being able to find System.Runtime and System.Reflection (even though they are definitely there and work with other configurations). I remember that I saw warning about this when the runtime errors were happening.

If you have a chance, try causing the same exception again, but this time when you build the project, look through the build log towards the beginning to see if it says anything about missing either or those libraries. When I saw them yesterday, I was able to actually run the app but it would crash as soon as it tried launching App.cs.
